feat: allow to use custom validation rules for registration (#244)
codeigniter4/shield: docs/7 - customization.md

    }
}
```

## Custom validation rules

### Registration

> The shield has the following rules for registration:

Reviewer: ```suggestion
Shield has the following rules for registration:
```

```php
[
    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
    'password'         => 'required|strong_password',
    'password_confirm' => 'required|matches[password]',
];
```

> If you need a different set of rules for registration, you can specify them in `Validation` configuration. You can use the configuration in `app\Config\Validation` and specify registration rules like:

Reviewer: ```suggestion
If you need a different set of rules for registration, you can specify them in your `Validation` configuration (**app\Config\Validation.php**) like:
```

```php
//--------------------------------------------------------------------
// Rules
//--------------------------------------------------------------------
public $registration = [
    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
    'password'         => 'required|strong_password',
    'password_confirm' => 'required|matches[password]',
];
```

Or use `Registrar`. If you are not aware of the Registrar, you can read more about the class [here](https://codeigniter4.github.io/CodeIgniter4/general/configuration.html#registrars).

> 

Reviewer: ```suggestion
For example:

```

```php
/**

>  * Registers the validation.

Reviewer: ```suggestion
 * Registers validation rules for registration.
```

    }
}
```

## Custom validation rules

### Registration

Shield has the following rules for registration:

```php
[
    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
    'password'         => 'required|strong_password',
    'password_confirm' => 'required|matches[password]',
];
```

If you need a different set of rules for registration, you can specify them in your `Validation` configuration (**app\Config\Validation.php**) like:

```php
//--------------------------------------------------------------------
// Rules
//--------------------------------------------------------------------
public $registration = [
    'username'         => 'required|alpha_numeric_space|min_length[3]|is_unique[users.username]',
    'email'            => 'required|valid_email|is_unique[auth_identities.secret]',
    'password'         => 'required|strong_password',
    'password_confirm' => 'required|matches[password]',
];
```

> Or use `Registrar`. If you are not aware of the Registrar, you can read more about the class [here](https://codeigniter4.github.io/CodeIgniter4/general/configuration.html#registrars).

Reviewer 2: @MGatner I don't see the use case that devs use Registrar like this to their apps.

Registrar needs existing config properties. So if we use Registrar, we need to do:
1. set `Config\Validation::$registration`
2. add `Registrar` to override the property values

Author: I'm using it like this now. My validation rules are defined only in Registrar, so I don't need to touch Validation config at all.

Reviewer 2: Oh really? 
Then the `Registrar` injects properties dynamically even if there are no properties in the config class?

Dynamic properties will be deprecated.
https://wiki.php.net/rfc/deprecate_dynamic_properties

Reviewer 2: @jozefrebjak You are correct. Registrar can add new properties.
https://github.com/codeigniter4/CodeIgniter4/blob/fe1dbc22003fb93d1b57addd3c35fe484cf5aedb/tests/system/Config/BaseConfigTest.php#L252-L255

Reviewer 2: I've confirmed that will not work on PHP 8.2.
https://github.com/codeigniter4/CodeIgniter4/issues/6162

Author: @kenjis good to know. So we need to define it in Validation config only. What is a purpose of the Registrar then ? Just to override packages configs ? I liked how it works now.

Reviewer: To be clear, I think in this case the property should be added to **app/Config/Validation.php**; using Registrar is an second option for cases where that isn't available (like Module implementation, or Local namespace).
